src/workerd/api/sql-test.js

```diff
@@ -60,6 +60,86 @@ async function test(storage) {
     assert.equal(result[2]['value'], 3)
   }
 
+
+  // Test partial query ingestion
+  assert.deepEqual(sql.ingest(`SELECT 123; SELECT 456;    `), '    ')
+  assert.deepEqual(sql.ingest(`SELECT 123; SELECT 456;`), '')
+  assert.deepEqual(sql.ingest(`SELECT 123; SELECT 456`), ' SELECT 456')
+  assert.deepEqual(sql.ingest(`SELECT 123; SELECT 45`), ' SELECT 45')
+  assert.deepEqual(sql.ingest(`SELECT 123; SELECT 4`), ' SELECT 4')
+  assert.deepEqual(sql.ingest(`SELECT 123; SELECT `), ' SELECT ')
+  assert.deepEqual(sql.ingest(`SELECT 123; SELECT`), ' SELECT')
+  assert.deepEqual(sql.ingest(`SELECT 123; SELEC`), ' SELEC')
+  assert.deepEqual(sql.ingest(`SELECT 123; SELE`), ' SELE')
+  assert.deepEqual(sql.ingest(`SELECT 123; SEL`), ' SEL')
+  assert.deepEqual(sql.ingest(`SELECT 123; SE`), ' SE')
+  assert.deepEqual(sql.ingest(`SELECT 123; S`), ' S')
+  assert.deepEqual(sql.ingest(`SELECT 123; `), ' ')
+  assert.deepEqual(sql.ingest(`SELECT 123;`), '')
+  assert.deepEqual(sql.ingest(`SELECT 123`), 'SELECT 123')
+  assert.deepEqual(sql.ingest(`SELECT 12`), 'SELECT 12')
+  assert.deepEqual(sql.ingest(`SELECT 1`), 'SELECT 1')
+
+  // Exec throws with trailing comments
+  assert.throws(
+    () => sql.exec('SELECT 123; SELECT 456; -- trailing comment'),
+    /SQL code did not contain a statement/
+  )
+  // Ingest does not
+  assert.deepEqual(
+    sql.ingest(`SELECT 123; SELECT 456; -- trailing comment`),
+    ' -- trailing comment'
+  )
+
+  // Test execution of ingested queries by taking an input of 6 INSERT statements, that all
+  // add 6 rows of data, then splitting that into a bunch of chunks, then ingesting them all
+  {
+    sql.exec(`CREATE TABLE streaming(val TEXT);`)
+
+    // Convert to binary otherwise .split can cause corruption for multi-byte chars
+    const inputBytes = new TextEncoder().encode(INSERT_36_ROWS)
+    const decoder = new TextDecoder()
+
+    // Use a chunk size 1, 3, 9, 27, 81, ... bytes
+    for (let length = 1; length < inputBytes.length; length = length * 3) {
+      let buffer = ''
+      for (let offset = 0; offset < inputBytes.length; offset += length) {
+        // Simulate a single "chunk" arriving
+        const chunk = inputBytes.slice(offset, offset + length)
+
+        // Append the new chunk to the existing buffer
+        buffer += decoder.decode(chunk, { stream: true })
+
+        // Ingest any complete statements and snip those chars off the buffer
+        buffer = sql.ingest(buffer)
```

Comment:
Nit: I wonder if it would be reasonable to have `sql.ingest(...)` also optionally accept the `TypedArray` here and perform the equivalent to `decoder.decode(...)` internally. It would save a bit of boilerplate in this pattern.

Reply:
Potentially, and this would avoid copying strings around, but I think what we have is good enough for a v1. Imo it'd be more likely to introduce something like `response.body.pipeTo(sql.ingester())` instead...